### PR TITLE
Add Mako support to do_not_disturb, usefull for wayland environment.

### PR DESCRIPTION
**Pull request by cyrinux**
Hi guys,

Here a tiny add, mako support for do do_not_disturb.

@maximbaz can you try ?

Regards,

### CODE REVIEW

File: py3status/modules/do_not_disturb.py [hunk 1]
```diff
@@ -102,6 +103,20 @@ def toggle(self, state):
...
 
+class Mako(Notification):
+    """
+    Mako Notification.
+    """
+
+    def toggle(self, state):
+        if state:
+            # pause
+            self.parent.py3.command_run("makoctl set invisible=1")
```
Comment (lasers) [inline py3status/modules/do_not_disturb.py]: (Alternative) Dunno if it's too long. Here's a line or two.
```python
self.parent.py3.command_run("makoctl set invisible={}".format(int(bool(state))))
```

File: py3status/modules/do_not_disturb.py [hunk 2]
```diff
@@ -56,6 +56,7 @@
 }
 ```
 
+@author Cyrinux https://github.com/cyrinux (mako)
```
Comment (lasers) [inline py3status/modules/do_not_disturb.py]: (Imho) Add contributors at end / bottom... following contribution timeline.

File: py3status/modules/do_not_disturb.py [hunk 3]
```diff
@@ -10,7 +10,7 @@
     pause: specify whether to pause or kill processes; for dunst
         see `Dunst Miscellaneous` section for more information
         (default False)
-    server: specify server to use, eg dunst or xfce4-notifyd, otherwise auto
+    server: specify server to use, eg dunst, xfce4-notifyd or mako, otherwise auto
```
Comment (lasers) [inline py3status/modules/do_not_disturb.py]: If you have `mako` and `dunst` installed, but are on `sway`, this would pick `dunst`. I wonder if mako should be checked first before dunst. `xfce4-notifyd` (or any other non-i3/sway notification servers) should always be last anyway.

File: py3status/modules/do_not_disturb.py [hunk 4]
```diff
@@ -135,7 +150,7 @@ class Py3status:
     thresholds = [(0, "bad"), (1, "good")]
 
     def post_config_hook(self):
-        servers = ["xfce4-notifyd", "dunst", None]
+        servers = ["xfce4-notifyd", "dunst", "mako", None]
```
Comment (lasers) [inline py3status/modules/do_not_disturb.py]: If you have `mako` and `dunst` installed, but are on `sway`, this would pick `dunst`. I wonder if mako should be checked first before dunst. `xfce4-notifyd` (or any other non-i3/sway notification servers) should always be last anyway.

(Thought) I wonder if `mako` would work on `i3`. If not, I could imagine some users switching between i3/sway sometimes to test py3status or other things... resulting in wrong server (i.e  `mako` on i3). I wonder if we need to check for `sway` too to see if we ought to run `mako` or `dunst`. Maybe it's an overkill. I don't know.

Comment (maximbaz) [reply py3status/modules/do_not_disturb.py]: `mako` doesn't work on i3, it was deliberate decision to prefer i3, as it is quite common for people to have both working i3 and sway in experimental mode, but once people switch to sway they no longer keep i3. I think it's fine like this (except your point that I agree with, we should move `mako` before `xfce4-notifyd`), defaults are defaults but users can control what they want in the config.

File: py3status/modules/do_not_disturb.py (hunk 2 above)
Comment (cyrinux) [reply py3status/modules/do_not_disturb.py]: My apologize. :)

File: py3status/modules/do_not_disturb.py (hunk 1 above)
Comment (cyrinux) [reply py3status/modules/do_not_disturb.py]: Neat, I keep this one.

File: py3status/modules/do_not_disturb.py [hunk 5]
```diff
@@ -109,12 +109,7 @@ class Mako(Notification):
     """
 
     def toggle(self, state):
-        if state:
-            # pause
-            self.parent.py3.command_run("makoctl set invisible=1")
-        else:
-            # resume
-            self.parent.py3.command_run("makoctl set invisible=0")
+        self.parent.py3.command_run("makoctl set invisible={}".format(int(bool(state))))
```
Comment (lasers) [inline py3status/modules/do_not_disturb.py]: Hmm. You don't need `bool()` if state is already bool. ;-)

File: py3status/modules/do_not_disturb.py [hunk 6]
```diff
@@ -145,7 +145,7 @@ class Py3status:
     thresholds = [(0, "bad"), (1, "good")]
 
     def post_config_hook(self):
-        servers = ["xfce4-notifyd", "dunst", "mako", None]
+        servers = ["dunst", "mako", "xfce4-notifyd", None]
```
Comment (lasers) [inline py3status/modules/do_not_disturb.py]: `servers = ["xfce4-notifyd", "mako", "dunst", "mako"]` due to `pgrep -f server`.

File: py3status/modules/do_not_disturb.py (hunk 3 above)
Comment (lasers) [inline py3status/modules/do_not_disturb.py]: This should say `mako, dunst, xfce4-notifyd`. Order of servers kinda matter too. Sorry for too many reviews.

File: py3status/modules/do_not_disturb.py (hunk 4 above)
Comment (lasers) [reply py3status/modules/do_not_disturb.py]: It is trivial to check for `sway` on `post_config_hook` to see if we need to `servers.insert(1, "mako")` before proceeding. Otherwise, this can auto pick mako on i3 if it is installed. 

>defaults are defaults but ...

Defaults are useful if they are smart enough to know which servers are supported.

>... but users can control what they want in the config.

Fine. As long as you acknowledged this can happen... Telling all users to change / control / fix their config instead of us auto detecting things like this in `post_config_hook` seems meh.

Carry on regardless. :+1:

File: py3status/modules/do_not_disturb.py (hunk 6 above)
Comment (maximbaz) [reply py3status/modules/do_not_disturb.py]: can you explain please? we will try each of them in order and stop on the first match, no? and we want dunst to be tried first, and `xfce4-notifyd` last?

I also don't understand why we have `None` in the list...

File: py3status/modules/do_not_disturb.py (hunk 3 above)
Comment (maximbaz) [reply py3status/modules/do_not_disturb.py]: We thank you for your reviews, do not apologize :wink: Agree that consistency matters :+1:

File: py3status/modules/do_not_disturb.py (hunk 6 above)
Comment (lasers) [reply py3status/modules/do_not_disturb.py]: Don't remember. I looked at module.

`None` is there because of `for server in servers` + `break` on L140-145 and maybe for initializing a failed backend due to all missing servers. Can't be sure.

IIRC `pgrep -f server` is required to detect running `xfce4-notifyd`. Otherwise, it will force running `dunst` instead of `xfce4-notifyd`. Also, if `dunst` is running, we sometimes can detect that too. If nothing, we go with `dunst` anyway... so `None` is there for this one too.

Comment (maximbaz) [reply py3status/modules/do_not_disturb.py]: @ricci hey, are you still using `xfce4-notifyd` and `py3status`? Would you be able to test something for us?

I'm pretty sure we don't need `-f` flag for `pgrep` as it can lead to false positives. Could you please try `do_not_disturb` module where this line is set to

    def post_config_hook(self):
        servers = ["dunst", "mako", "xfce4-notifyd"]

And also could you please post the result of these commands?

```
$ pgrep -fl xfce4-notifyd
$ pgrep -fa xfce4-notifyd
$ pgrep -xl xfce4-notifyd
```

Basically, to make the precise search work, we must do `pgrep -x <result of pgrep -fl xfce4-notifyd>`, that's what I'd like to figure out.

Comment (ricci) [reply py3status/modules/do_not_disturb.py]: Yes, the latest version of the code in the pull request works for me with `xfce4-notifyd`. Here are the outputs of those commands:
```
> pgrep -fl xfce4-notifyd 
749 xfce4-notifyd
> pgrep -fa xfce4-notifyd
749 /usr/lib/xfce4/notifyd/xfce4-notifyd
 > pgrep -xl xfce4-notifyd
749 xfce4-notifyd
```
I will note that I'm running `xfce4-notifyd` via `systemcl --user start xfce4-notifyd.service`, which does start the process and keep it running continually. There may be some other way to set it to launch automatically via a socket or `dbus`, meaning that the process might not always be running (I don't know).

Comment (maximbaz) [reply py3status/modules/do_not_disturb.py]: Great, thank you very much! 

I think we should switch to the precise `pgrep -x` from `pgrep -f`, as the last test above shows that it works for `xfce4-notifyd`, and I think we need to set `servers = ["dunst", "mako", "xfce4-notifyd"]`

**Comment by ultrabug** [comment]
Thanks!
